### PR TITLE
Refactor Catalog API

### LINKED ISSUE #1929: FE support describing schemas using JsonSchema
## Tell us about the problem you're trying to solve
Currently the FE relies on a data model that assumes a flat data schema (table / column). We need to allow nesting. For this we will describe schemas using the full power of JsonSchema.

For this issue the focus is to just get the API to work exactly as it is now (no changes to the visuals of the UI), while using the new data model. We discussed offline and agreed to go with option 1 as described in this [PR](https://github.com/airbytehq/airbyte/pull/1924/files). To be clear we will pass nested data to the UI, the nested fields will not be displayed in the UI (we can worry about that later), but they _will_ be passed back.

@ChristopheDuong  is working on implementing that interface in the API.

Relates to https://github.com/airbytehq/airbyte/issues/1491

This is our highest priority for the week.

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
Closes #1929 

## How
Implements option1 of #1924:

The intermediate layer StandardDataSchema in the back-end is now removed, we can directly convert an `AirbyteCatalog` object from the Airbyte Protocol into a new `AirbyteCatalog` object generated from the API that reproduces almost the same fields/structures as the one from the Airbyte Protocol.

The main difference is that the API doesn't make a distinction between an `AirbyteCatalog` and a `ConfiguredAirbyteCatalog`: The API `AirbyteCatalog` has a default `AirbyteStreamConfiguration` for each stream included.

The new big change is that instead of storing an array of Fields, we now manipulate the JSON Schema objects directly, and thus, it can represent complex structures such as nested schema and arrays, etc

Note that:
- we also lose the `dataType`, `cleanedName`, and `selected` fields for each fields by doing so. (could be added back inside the Json schema itself?)
- we keep the `cleanedName`, and `selected` fields at the stream level though

## Pre-merge Checklist
- [X] *Run integration/Acceptance tests*
- [ ] *Adapt FE to new API*

## Recommended reading order

1. `airbyte-api/src/main/openapi/config.yaml`
2. `airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java`
3. the rest

### CODE REVIEW

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java [hunk 1]
```diff
@@ -24,107 +24,87 @@
...
 
-    return new Schema().withStreams(persistenceStreams);
+  public static io.airbyte.api.model.AirbyteCatalog convertTo(final io.airbyte.protocol.models.AirbyteCatalog catalog) {
+    return new io.airbyte.api.model.AirbyteCatalog()
+        .streams(catalog.getStreams()
+            .stream()
+            .map(SchemaConverter::convertTo)
+            .map(s -> new io.airbyte.api.model.AirbyteStreamAndConfiguration()
+                .stream(s)
+                ._configuration(getDefaultAirbyteStreamConfiguration(s)))
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:64]: why is this underscored?

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java [hunk 2]
```diff
@@ -24,107 +24,87 @@
...
-      field.withDataType(Enums.convertTo(apiField.getDataType(), DataType.class));
-    } else {
-      field.withDataType(DataType.OBJECT);
-    }
-    return field;
+  private static io.airbyte.api.model.AirbyteStreamConfiguration getDefaultAirbyteStreamConfiguration(io.airbyte.api.model.AirbyteStream s) {
+    io.airbyte.api.model.AirbyteStreamConfiguration result = new AirbyteStreamConfiguration()
+        .cleanedName(Names.toAlphanumericAndUnderscore(s.getName()))
+        .cursorField(s.getDefaultCursorField())
+        .selected(true);
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:72]: i don't like that these converters still aren't symmetrical. this is a fault in the BE data model (not in your implementation). we still lose information as the configuration comes back from the API. can you create an issue to make this symmetrical? i think ideally the backend model should look more like the FE. keeping all of the streams in the list but having a flag as to whether they are going to be synced. it's a bigger change though and we should not try to do it in this PR, but want to make sure we follow up.

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java [hunk 3]
```diff
@@ -24,107 +24,87 @@
...
-  public static Schema toPersistenceSchema(SourceSchema sourceSchema) {
-    final List<Stream> persistenceStreams =
-        sourceSchema.getStreams()
-            .stream()
-            .map(apiStream -> {
-              final List<Field> persistenceFields = apiStream.getFields()
-                  .stream()
-                  .map(SchemaConverter::toPersistenceField)
-                  .collect(Collectors.toList());
+  public static io.airbyte.api.model.AirbyteStream convertTo(final io.airbyte.protocol.models.AirbyteStream stream) {
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:39]: i think these names need to include more context. we are converting back and forth between 2 models with the same name and then the other one converts in the other direction. this is hard to parse as a human if they are just called `convertTo`. it is super easy to mess up.

i would suggest `toApi` and to `toBackend`. Not married to these exact names, but we need something to tell us what "direction" the conversion is happening in.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java [hunk 4]
```diff
@@ -77,9 +77,9 @@ public ConnectionRead createConnection(ConnectionCreate connectionCreate) throws
         .withStatus(toPersistenceStatus(connectionCreate.getStatus()));
 
     if (connectionCreate.getSyncSchema() != null) {
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:79]: maybe worth adding a todo here that this behavior is probably not deseriable. i don't think sending a null configured catalog should be valid. not the right time to change this but let's leave ourselves a breadcrumb so we can come back to it.

File: airbyte-server/src/test/java/io/airbyte/server/converters/SchemaConverterTest.java [hunk 5]
```diff
@@ -26,59 +26,21 @@
...
-          .sourceDefinedCursor(false)
-          .supportedSyncModes(Lists.newArrayList(io.airbyte.api.model.SyncMode.FULL_REFRESH, io.airbyte.api.model.SyncMode.INCREMENTAL))
-          .defaultCursorField(Lists.newArrayList(COLUMN_ID))
-          .syncMode(io.airbyte.api.model.SyncMode.INCREMENTAL)
-          .cursorField(Lists.newArrayList(COLUMN_ID))));
-
   @Test
-  void convertToPersistenceSchema() {
-    assertEquals(SCHEMA, SchemaConverter.toPersistenceSchema(API_SCHEMA));
+  void convertToProtocol() {
```
Comment (cgardens) [inline airbyte-server/src/test/java/io/airbyte/server/converters/SchemaConverterTest.java:37]: by convention test names should include the word `test`

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java (hunk 3 above)
Comment (cgardens) [reply airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:39]: the names you use in your tests are also fine.

File: airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java [hunk 6]
```diff
@@ -82,33 +82,25 @@
...
+    final ConfiguredAirbyteCatalog catalog = new ConfiguredAirbyteCatalog().withStreams(Collections.singletonList(stream));
 
     final StandardSync standardSync = new StandardSync()
         .withConnectionId(connectionId)
         .withDestinationId(destinationId)
         .withSourceId(sourceId)
         .withStatus(StandardSync.Status.ACTIVE)
         .withName(CONNECTION_NAME)
-        .withSchema(schema);
+        .withSchema(catalog);
```
Comment (cgardens) [inline airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java:95]: this should probably be called catalog instead of schema. i'm sympathetic that changing these can be a bit of a pain, so if this is really awful to fix don't worry about it. but since we're tearing everythhing up now, this is as good a time as any.

**Review by cgardens — APPROVED** [review]
A few comments but I think they are mostly superficial. Will approve now, but we cannot merge this until corresponding FE changes are made.

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java (hunk 1 above)
Comment (ChristopheDuong) [reply airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:64]: This is how the code generator spat out, maybe because it's some sort of keyword?

I can change the name to avoid this

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java (hunk 3 above)
Comment (ChristopheDuong) [reply airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java:39]: Ok

File: airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java (hunk 6 above)
Comment (ChristopheDuong) [reply airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java:95]: Don't worry, it's not as painful as changing completely the object underneath haha

File: airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java (hunk 1 above)
Comment (cgardens) [reply airbyte-server/src/main/java/io/airbyte/server/converters/SchemaConverter.java]: nah. it's okay. as long as it is an intentional thing to avoid a namespace confusion i think it's okay.

File: airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java (hunk 6 above)
Comment (cgardens) [reply airbyte-workers/src/test/java/io/airbyte/workers/TestConfigHelpers.java]: I don't think it should affect him since this model is only used behind the API. the only thing that should affect him are changes to the API interface (in config.yaml).

File: airbyte-api/src/main/openapi/config.yaml [hunk 7]
```diff
@@ -1692,65 +1692,72 @@ components:
...
       required:
         - streams
       properties:
         streams:
           type: array
           items:
-            $ref: "#/components/schemas/SourceSchemaStream"
-    SourceSchemaStream:
+            $ref: "#/components/schemas/AirbyteStreamAndConfiguration"
+    AirbyteStreamAndConfiguration:
```
Comment (Jamakase) [inline airbyte-api/src/main/openapi/config.yaml:1706]: What's the point of prepending entities with `Airbyte`? Most of the other fields don't have this prefix. What is more, I do not think that it actually brings any meaning to prepend all fields with Airbyte

Comment (Jamakase) [reply airbyte-api/src/main/openapi/config.yaml:1706]: Also, @ChristopheDuong do you think its a good idea to combine different parts of this object with `And`? What if we add 1 more field here somewhere in future? It means that the current name will be obsolete.

Comment (cgardens) [reply airbyte-api/src/main/openapi/config.yaml:1706]: Stream is a reserved word in java and so that's why we tend to stick the word Airbyte in front of it. We will need to come up with a better name for this this thing. The and is not ideal, but we just don't have a better name yet.
